### PR TITLE
TYP: use __all__ to signal public API to type checkers

### LINKED ISSUE #43664: TYP: Redundant imports to define the public API
The following is not standardized in a PEP but is [generally agreed on by multiple type checkers](https://github.com/pytorch/pytorch/issues/47027#issuecomment-780898127) and is mentioned in the [scope of stub files in PEP484](https://www.python.org/dev/peps/pep-0484/#stub-files). See https://github.com/pytorch/pytorch/issues/47027#issuecomment-719036333 for more pointers. Pytorch (in [__init__.py](https://github.com/pytorch/pytorch/blob/88032d894311e5c0aed8bbc21a4306bc6be4af82/torch/__init__.py#L678)) and numpy (in [__init__.pyi](https://github.com/numpy/numpy/blob/df0b1bd75f0e8489912863d1f9dfde1846a7bf51/numpy/__init__.pyi#L198)) are two of the projects that already try to use redundant imports.

#### Is your feature request related to a problem?

When pandas becomes a py.typed library (#28142), static type checks need to know what is part of the public API and what isn't. PEP 484 outlines that imported modules are only part of the public API, if they follow the pattern of "import X as X" (for stub files). Pyright implemented this not only for stub files but also for "normal" python code (see https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface).
Edit: mypy also complains about that when `implicit_reexport=false`

#### Describe the solution you'd like

Use redundant imports to signal what is part of the public API.

While this is not standardized for normal python code, pandas has/will have many `.pyi` files for its cython code (which is subject to PEP 484). It is probably more consistent to use redundant imports in the entire codebase, not only in `.pyi` files.

#### API breaking implications

Until pandas is a py.typed library, it would be good to have a clear statement what takes precedence to define the public API (in case there are mismatches between the documentation, import statements, or type annotations).

#### Additional context

When creating an empty py.typed file in the pandas directory and using pyright to validate the following code:

```python
import pandas as pd

pd.DataFrame()
```
`error: "DataFrame" is not exported from module "pandas" (reportPrivateImportUsage)`

### PR DESCRIPTION
**Pull request by twoertwein**
- [x] closes #43664
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/__init__.py [hunk 1]
```diff
@@ -31,150 +31,153 @@
...
+# public API. Pandas is not (yet) a py.typed library: the public API is determined
+# based on the documentation.
 from pandas._config import (
-    get_option,
-    set_option,
-    reset_option,
-    describe_option,
-    option_context,
-    options,
+    get_option as get_option,
```
Comment (jreback) [inline pandas/__init__.py:38]: wouldn't it be better to simply have a `__all__` ? (i think we actually test the visibility of the names already)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
cc @simonjayhawkins

**Review by simonjayhawkins — APPROVED** [review]
LGTM ex @jreback comment. This is a precursor to #43672?

**Comment by twoertwein** [comment]
> This is a precursor to #43672?

No, but once we add a py.typed file we could use pyright to make sure that a list of intended to be public functions/attributes are picked up as public by type checkers.

File: pandas/__init__.py (hunk 1 above)
Comment (twoertwein) [reply pandas/__init__.py:38]: Yes, that is possible!

https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md#library-interface
> A module can expose an __all__ symbol at the module level that provides a list of names that are considered part of the interface. This overrides all other rules above, allowing imported symbols or symbols whose names begin with an underscore to be included in the interface.

File: pandas/__init__.py [hunk 2]
```diff
@@ -170,8 +170,8 @@
 from pandas.io.json import _json_normalize as json_normalize
 
 from pandas.util._tester import test
-import pandas.testing
-import pandas.arrays
+from pandas import testing
+from pandas import arrays
```
Comment (twoertwein) [inline pandas/__init__.py:174]: Is there a reason why `api`, `testing`, and `arrays` was not imported with "from pandas import"? I changed it to `from pandas import` to include them in `__all__`.

File: pandas/__init__.py (hunk 1 above)
Comment (twoertwein) [reply pandas/__init__.py]: Replaced the `X as X` with a long `__all__`.

Comment (twoertwein) [reply pandas/__init__.py]: Should `pandas.core.config_init` be added to `__all__`? I didn't add it
https://github.com/pandas-dev/pandas/blob/9e9c0de285777453eab68069c1292d0ba4e23f84/pandas/__init__.py#L43

Comment (jreback) [reply pandas/__init__.py]: umm if you don't need it then no

File: pandas/__init__.py (hunk 2 above)
Comment (jreback) [reply pandas/__init__.py:174]: these should be equivalent (ex the name space visiblity).

note that we may not be fully testing this (see in pandas/tests/api/test_api.py)

File: pandas/__init__.py [hunk 3]
```diff
@@ -303,3 +302,118 @@ def __getattr__(name):
   - Time series-specific functionality: date range generation and frequency
     conversion, moving window statistics, date shifting and lagging.
 """
+
+# Use __all__ to let type checkers know what is part of the public API.
+# Pandas is not (yet) a py.typed library: the public API is determined
+# based on the documentation.
+__all__ = [
```
Comment (jreback) [inline pandas/__init__.py:309]: can you alphabetize?

Comment (twoertwein) [reply pandas/__init__.py:309]: sure, I kept them in the order they were imported.

Comment (twoertwein) [reply pandas/__init__.py:309]: sorted

File: pandas/__init__.py (hunk 1 above)
Comment (twoertwein) [reply pandas/__init__.py]: The existing tests in pandas/tests/api/test_api.py do not expect `config_init` to be exported.

File: pandas/__init__.py (hunk 2 above)
Comment (twoertwein) [reply pandas/__init__.py:174]: I added a test for `__all__` and found in that process a few un-exported modules.

File: pandas/tests/api/test_api.py [hunk 4]
```diff
@@ -217,6 +217,26 @@ def test_api(self):
...
+            + self.funcs_json
+            + self.funcs_to
+        )
+        actual = set(pd.__all__)
+
+        extraneous = actual - expected
+        assert not extraneous
+
+        missing = expected - actual
+        assert not missing
```
Comment (twoertwein) [inline pandas/tests/api/test_api.py:238]: `assert actual == expected` would work, but it doesn't create a nice log in pytest.

**Comment by twoertwein** [comment]
I think the modules I added in the last commit are not meant to be public (compat, core, io, ...)? @simonjayhawkins 

If the above modules are not meant to be public, there probably needs to be a discussion which additional classes should be exported in `pandas/__init__.py`. I think it would currently not be possible for a user to do

```py
def mean(grouper: pd.core.groupby.DataFrameGroupBy) -> pd.DataFrame:
    return group.mean()

```
without using "private" APIs (pd.core.groupby.DataFrameGroupBy).

**Comment by jreback** [comment]
none of those are documented as public 

we have an issue about deprecating and making these private but hasn't gotten done

don't add them explicitly to any exports

**Comment by jreback** [comment]
some of these apis should be exported in pd.api as well (eg DataframeGroupby for example)

**Comment by twoertwein** [comment]
So the only public "top-level sub-packages" are: "api", "arrays", "options", "test", "testing"? They are all imported in `pandas/__init__.py`.

**Comment by twoertwein** [comment]
Reading through the [documentation](https://pandas.pydata.org/docs/reference/index.html), the following should be listed as public in `pandas/__init__.py`:
pandas.errors: just defines warnings/errors
pandas.plotting: has itself `__all__` to define what is public
pandas/testing.py: has itself `__all__` to define what is public
pandas/io: contains some public classes but is missing an `__all__`
pandas/tseries: contains some public classes but is missing an `__all__`

**Comment by twoertwein** [comment]
I was trying to expose only meant-to-be-public sub-modules (public = [listed](https://pandas.pydata.org/docs/search.html) in the documentation) in `pandas/io/__init__.py` and `pandas/tseries/__init__.py` by importing them and adding them to `__all__` (and then repeating this for their sub-modules that do not yet have an `__all__`). Unfortunately, this leads to circular imports :(

Here is an example how I would `pandas/io/__init__.py` to look like
```py
# import modules that have public classes/functions
from pandas.io import (
    formats,
    json,
    stata,
)

# and mark only those modules as public
__all__ = ["formats", "json", "stata"]
```

I think many people assume that everything that doesn't start with an underscore is public. `__all__` could be a gentle nudge to remind people what is public. For example, after renaming `get_filepath_buffer` to `_get_filepath_buffer` in #37639 it is clear that at least two projects (https://github.com/PEtab-dev/PEtab/pull/493 and https://github.com/fphammerle/freesurfer-stats/pull/15) assumed it to be public even though it is not listed in the documentation.

**Comment by jreback** [comment]
can you rebase

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/pull/43695/checks?check_run_id=3734385197

hmm maybe need to be more explict in the docs

**Comment by twoertwein** [comment]
> https://github.com/pandas-dev/pandas/pull/43695/checks?check_run_id=3734385197

`from pandas import *` causes the deprecated models to be imported as they are listed in `__all__`.

I think there are two options:

1. don't use `from pandas import *`
2. do not add the deprecated modules to `__all__`

Which one is preferred?

**Comment by jreback** [comment]
i don't think we should add the deprecated modules to all

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @twoertwein very nice!

**Pull request merged**